libbeat/outputs/shipper/shipper.go

```diff
@@ -153,37 +172,65 @@ func (c *shipper) Publish(ctx context.Context, batch publisher.Batch) error {
 
 	ctx, cancel := context.WithTimeout(ctx, c.timeout)
 	defer cancel()
-	resp, err := c.client.PublishEvents(ctx, &messages.PublishRequest{
+	publishReply, err := c.client.PublishEvents(ctx, &messages.PublishRequest{
+		Uuid:   c.serverID,
 		Events: convertedEvents,
 	})
 
-	if status.Code(err) != codes.OK || resp == nil {
+	if status.Code(err) != codes.OK {
 		batch.Cancelled()         // does not decrease the TTL
 		st.Cancelled(len(events)) // we cancel the whole batch not just non-dropped events
 		return fmt.Errorf("failed to publish the batch to the shipper, none of the %d events were accepted: %w", len(convertedEvents), err)
 	}
 
 	// with a correct server implementation should never happen, this error is not recoverable
-	if int(resp.AcceptedCount) > len(nonDroppedEvents) {
+	if int(publishReply.AcceptedCount) > len(nonDroppedEvents) {
 		return fmt.Errorf(
 			"server returned unexpected results, expected maximum accepted items %d, got %d",
 			len(nonDroppedEvents),
-			resp.AcceptedCount,
+			publishReply.AcceptedCount,
 		)
 	}
 
-	// the server is supposed to retain the order of the initial events in the response
-	// judging by the size of the result list we can determine what part of the initial
-	// list was accepted and we can send the rest of the list for a retry
-	retries := nonDroppedEvents[resp.AcceptedCount:]
+	ackClient, err := c.client.PersistedIndex(ctx, &messages.PersistedIndexRequest{
+		PollingInterval: durationpb.New(c.config.AckPollingInterval),
+	})
+	if err != nil {
+		return fmt.Errorf("acknowledgement failed due to the connectivity error: %w", err)
+	}
+
+	for {
+		indexReply, err := ackClient.Recv()
+		if err != nil {
+			return fmt.Errorf("acknowledgement failed due to the connectivity error: %w", err)
+		}
+
+		if indexReply.GetUuid() != c.serverID {
+			batch.Cancelled()
+			st.Cancelled(len(events))
+			err := fmt.Errorf("acknowledgement failed due to a connection to a different server %s, expected %s", indexReply.Uuid, c.serverID)
+			c.serverID = indexReply.GetUuid()
```

Comment:
Following on the other comment about `serverID` initialization, this `serverID` update has no effect since it will be immediately overwritten by a reconnect

Reply:
How about moving this `if` block out of the loop with the first `Recv` call so we only do this once? This is all one RPC call so the server ID can't change between iterations.

Reply:
@faec well, as you said this line has no effect because of the pipeline logic which is supposed to call `Connect` again. Why moving this to somewhere else if we can remove?

Also, none of this logic and edge cases are final. We'll find out much more after the integration tests are done.

Reply:
For removal I was talking about `serverID` but the moving was the `if` block, `if indexReply.GetUuid() != c.serverID`, since I assume we do want to check server id once but there's no need to check it multiple times for a stable connection -- if the first read is before the infinite loop, then the rest of the loop can do it the simple way.

Reply:
@faec I moved the check in https://github.com/elastic/beats/pull/32708/commits/f6795c3ff42c906d7ed763ff73a61b43f8ff5b15

I also realised that `CloseSend` is not the right way to close the stream, the gRPC doc requests to cancel the context, so I did that too.